Sources/SourceKitLSP/Swift/SyntacticTestIndex.swift

```diff
@@ -0,0 +1,229 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import LSPLogging
+import LanguageServerProtocol
+import SKSupport
+
+/// Task metadata for `SyntacticTestIndexer.indexingQueue`
+fileprivate enum TaskMetadata: DependencyTracker, Equatable {
+  case read
+  case index(Set<DocumentURI>)
+
+  /// Reads can be concurrent and files can be indexed concurrently. But we need to wait for all files to finish
+  /// indexing before reading the index.
+  func isDependency(of other: TaskMetadata) -> Bool {
+    switch (self, other) {
+    case (.read, .read):
+      // We allow concurrent reads
+      return false
+    case (.read, .index(_)):
+      // We allow index tasks scheduled after a read task to be be executed before the read.
+      // This effectively means that a `read` requires the index to be updated *at least* up to the state at which the
+      // read was scheduled. If more changes come in in the meantime, it is OK for the read to pick them up. This also
+      // ensures that reads aren't parallelization barriers.
+      return false
+    case (.index(_), .read):
+      // We require all index tasks scheduled before the read to be finished.
+      // This ensures that the index has been updated at least to the state of file at which the read was scheduled.
+      // Adding the dependency also elevates the index task's priorities.
+      return true
+    case (.index(let lhsUris), .index(let rhsUris)):
+      // Technically, we should be able to allow simultaneous indexing of the same file. When a file gets re-scheduled
+      // for indexing, all previous index invocations should get cancelled. But conceptually the code becomes simpler
```

Comment:
> all previous index invocations should get cancelled

Isn't true any more. Doesn't really matter, it's just the comment is out of date.